src/core/python2js_buffer.js

```diff
@@ -0,0 +1,214 @@
+        // clang-format on
+        throw new Error("BigInt64Array is not supported on this browser." +
+                        errorMessage);
+      }
+      arrayType = BigInt64Array;
+      break;
+    case 'Q':
+      // clang-format off
+            if (globalThis.BigUint64Array === undefined) {
+        // clang-format on
+        throw new Error("BigUint64Array is not supported on this browser." +
+                        errorMessage);
+      }
+      arrayType = BigUint64Array;
+      break;
+    case 'f':
+      arrayType = Float32Array;
+      break;
+    case 'd':
+      arrayType = Float64Array;
+      break;
+    case "e":
+      throw new Error("Javascript has no Float16 support.");
+    default:
+      throw new Error(`Unrecognized format character '${formatChar}'.` +
+                      errorMessage);
+    }
+    return [ arrayType, bigEndian ];
+  };
+
+  Module.python2js_buffer_1d_contiguous = function(ptr, stride, n, converter) {
+    "use strict";
+    let byteLength = stride * n;
+    let backing = HEAP8.slice(ptr, ptr + byteLength).buffer;
+    return converter(backing);
+  };
+
+  Module.python2js_buffer_1d_noncontiguous = function(ptr, stride, suboffset, n,
+                                                      itemsize, converter) {
+    "use strict";
+    let byteLength = itemsize * n;
+    let buffer = new Uint8Array(byteLength);
+    for (i = 0; i < n; ++i) {
+      let curptr = ptr + i * stride;
+      if (suboffset >= 0) {
+        curptr = HEAP32[curptr / 4] + suboffset;
+      }
+      buffer.set(HEAP8.subarray(curptr, curptr + itemsize), i * itemsize);
+    }
+    return converter(buffer.buffer);
+  };
+
+  Module._python2js_buffer_recursive = function(ptr, curdim, bufferData) {
+    "use strict";
+    let n = HEAP32[bufferData.shape / 4 + curdim];
+    let stride = HEAP32[bufferData.strides / 4 + curdim];
+    let suboffset = -1;
+    // clang-format off
+    if (bufferData.suboffsets !== 0) {
+      suboffset = HEAP32[bufferData.suboffsets / 4 + curdim];
+    }
+    if (curdim === bufferData.ndim - 1) {
+      if (stride === bufferData.itemsize && suboffset < 0) {
+        // clang-format on
+        return Module.python2js_buffer_1d_contiguous(ptr, stride, n,
+                                                     bufferData.converter);
+      } else {
+        return Module.python2js_buffer_1d_noncontiguous(ptr, stride, suboffset,
+                                                        n, bufferData.itemsize,
+                                                        bufferData.converter);
+      }
+    }
+
+    let result = [];
+    for (let i = 0; i < n; ++i) {
+      let curPtr = ptr + i * stride;
+      if (suboffset >= 0) {
+        curptr = HEAP32[curptr / 4] + suboffset;
+      }
+      result.push(
+          Module._python2js_buffer_recursive(curPtr, curdim + 1, bufferData));
+    }
+    return result;
+  };
+
+  Module.get_converter = function(format, itemsize) {
```

Comment:
Could you please add a docstring here and for all top level functions in this file?